### DISCUSSION TITLE
[STM32] DWC2 crashes when configuring IRQS

### DISCUSSION THREAD

**Discussion by hitech95** [hathach/tinyusb#2115] [Q&A]
Hi,
I'm trying to use this library with the stm32duino core and platformio.
The board is a Black pill F401CC.

That said I want to start to experiment with HID IN/OUT and in future to have multiple endpoint?
I still have to read all the stuff in the USB specs.

That said, I've imported the library as a submodule. but I had to apply some changes due to some conficts between the old synaptic and new dwc2 driver.

Now I can properly build the code but when calling `tud_init(BOARD_TUD_RHPORT)` the board freezes.

I'm not 100% sure on why but I've seen that
 it might be due to some IRQ issue.
If I comment out the following section
```c
  // Required as part of core initialization.
  // TODO: How should mode mismatch be handled? It will cause
  // the core to stop working/require reset.
  dwc2->gintmsk = GINTMSK_OTGINT   | GINTMSK_MMISM  | GINTMSK_RXFLVLM  |
                  GINTMSK_USBSUSPM | GINTMSK_USBRST | GINTMSK_ENUMDNEM | GINTMSK_WUIM;
```
https://github.com/hathach/tinyusb/blob/master/src/portable/synopsys/dwc2/dcd_dwc2.c#L540

The board doesn't lock.

My code is basically:

```c
#include <Arduino.h>
#include "tusb.h"

// Pinmux for USB not exposed by core variants !
const PinMap PinMap_USB_OTG_FS[] = {
    {PA_11, USB_OTG_FS, STM_PIN_DATA(STM_MODE_AF_PP, GPIO_PULLUP, GPIO_AF10_OTG_FS)}, // USB_OTG_FS_DM
    {PA_12, USB_OTG_FS, STM_PIN_DATA(STM_MODE_AF_PP, GPIO_PULLUP, GPIO_AF10_OTG_FS)}, // USB_OTG_FS_DP
    {NC, NP, 0}
};

/**
  * @brief  This function handles USB-On-The-Go FS/HS global interrupt request.
  * @param  None
  * @retval None
  */
#ifdef USE_USB_HS
void OTG_HS_IRQHandler(void)
#elif defined(USB_OTG_FS)
void OTG_FS_IRQHandler(void)
#else /* USB */
void USB_IRQHandler(void)
#endif
{
  tud_int_handler(0);
}

static void setupTinyUSB()
{
  const PinMap *map = NULL;

  // set pin mapping (using core library)
  map = PinMap_USB_OTG_FS;
  while (map->pin != NC)
  {
    pin_function(map->pin, map->function);
    map++;
  }
  
  /* Enable USB FS Clock */
  __HAL_RCC_USB_OTG_FS_CLK_ENABLE();
  
  // Blackpill doesn't use VBUS sense (B device) explicitly disable it
  USB_OTG_FS->GCCFG |= USB_OTG_GCCFG_NOVBUSSENS;
  USB_OTG_FS->GCCFG &= ~USB_OTG_GCCFG_VBUSBSEN;
  USB_OTG_FS->GCCFG &= ~USB_OTG_GCCFG_VBUSASEN;
  
  tud_init(BOARD_TUD_RHPORT);
}
  
void setup()
{
  setupTinyUSB();
  printf("HelloWorld!")
}

void loop()
{
  tud_task(); // tinyusb device task
 }
 ```
 
Any suggestion?
Yes I've attached the logs below and yes there are the calbacks in the main.cpp file.
My descriptor and config are taken from the https://github.com/hathach/tinyusb/tree/master/examples/device/hid_generic_inout.

[log_tinyusb.txt](https://github.com/hathach/tinyusb/files/11775145/log_tinyusb.txt)

**Comment by manuelbl** [comment]
Since you include `Arduino.h`, this is likely C++ code and not C code. So there is a chance that your IRQ handler is not called because of invalid linkage. If so, a default IRQ handlers implementation stopping the program will be called instead.

Try to add `extern "C"` to the IRQ handlers, e.g.:

```
extern "C" void OTG_FS_IRQHandler(void)
{
  tud_int_handler(0);
}
```

The "C" linkage can also be specified in a declaration, possibly in a header file. But from the files you include, this is unlikely.

**Reply by hitech95** [reply]
Thanks this saved me, it was a dumb mistake but I have completly missed it!

**Comment by hathach** [comment]
Blackpill F401CC is one of supported boards. You can try to run with stock exmaple to verify that it works then use that as base to compare your code with.

**Reply by hitech95** [reply]
I have a question that I've not seen anywhere how can I select on wat class to output an HID report to?

For example I have my device to be composite type, two HID classes. the first one has only a single HID decriptior while the second interface is also HID composite and has 3 descriptors:
```c

uint8_t const desc_hid_report_keyboard_simple[] =
    {
        TUD_HID_REPORT_DESC_KEYBOARD()
        };

uint8_t const desc_hid_report_keyboard[] =
    {
        TUD_HID_REPORT_DESC_KEYBOARD(HID_REPORT_ID(REPORT_ID_KEYBOARD)),
        TUD_HID_REPORT_DESC_CONSUMER(HID_REPORT_ID(REPORT_ID_CONSUMER_CONTROL)),
        TUD_HID_REPORT_DESC_GENERIC_INOUT(CFG_TUD_HID_EP_BUFSIZE, HID_REPORT_ID(REPORT_ID_RGB)),
    };
    
uint8_t const desc_configuration[] =
    {
        // Config number, interface count, string index, total length, attribute, power in mA
        TUD_CONFIG_DESCRIPTOR(1, ITF_NUM_TOTAL, 0, CONFIG_TOTAL_LEN, 0x00, 100),

        // Interface number, string index, protocol, report descriptor len, EP Out & In address, size & polling interval
        TUD_HID_INOUT_DESCRIPTOR(ITF_NUM_HID_SPL, 4, HID_ITF_PROTOCOL_NONE, sizeof(desc_hid_report_keyboard_simple), EPNUM_HID_MOUT, EPBASE_HID_MIN, CFG_TUD_HID_EP_BUFSIZE, 100),

        // Interface number, string index, protocol, report descriptor len, EP Out & In address, size & polling interval
        TUD_HID_INOUT_DESCRIPTOR(ITF_NUM_HID_ADV, 5, HID_ITF_PROTOCOL_NONE, sizeof(desc_hid_report_keyboard_advanced), EPNUM_HID_KOUT, EPBASE_HID_KIN, CFG_TUD_HID_EP_BUFSIZE, 100),
    };
 ```
 
 to emulate a key press I can call and pass the instance number:
 ```c
 tud_hid_keyboard_report(INTERFACE_HID1, REPORT_ID_KEYBOARD, 0, keycode);
 ```
I suppose that if I want to communicate with the firtst interface I should use `INTERFACE_HID0` and `0` as the `report_id`.

Now my question is how can I identify the single HID report ID on the host side? I want to try to read/write to the REPORT_ID_RGB of instance 1. Do you have any info about this process to point me? I've looked in the HID specs but I didnt find it!

Another issue I'm facing is that most of the time the key seem to be stuck down. And it keep writing `a` until I reset the MCU.
I' have the DEBUG set to 2 so I can see that on release the reports callbacks stops, so I suppose that the key release should have been transmitted correctly. any way to debug it? I thoung it was an issue with the button debounce but it doesnt seems to be. using a single HID keyboard seems to solve the issue.
